File: jmix-custom-table/src/main/java/com/company/customcomponent/bridge/ReactBridge.java
package com.company.customcomponent.bridge;

import com.vaadin.flow.component.ClientCallable;
import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;

import java.util.function.Consumer;

@Tag("react-bridge")
@JsModule("./react-bridge/react-bridge.ts")
public class ReactBridge extends Component {

    private Consumer<String> onMessage;

    public void setOnMessage(Consumer<String> onMessage) {
        this.onMessage = onMessage;
    }

    @ClientCallable
    private void onReactEvent(String json) {
        if (onMessage != null) onMessage.accept(json);
    }
}

File: jmix-custom-table/src/main/frontend/react-bridge/react-bridge.ts
declare global {
    interface Window {
        ReactBridge?: { notify: (msg: any) => void };
    }
}

function getBridgeEl(): any | null {
    // react-bridge là tag của component Java @Tag("react-bridge")
    return document.querySelector("react-bridge");
}

window.ReactBridge = {
    notify: (msg: any) => {
        const el = getBridgeEl();

        // Khi Vaadin chưa attach component, $server có thể chưa sẵn sàng
        if (!el?.$server?.onReactEvent) {
            console.warn("[ReactBridge] element not ready or missing $server");
            return;
        }

        // Gọi thẳng @ClientCallable onReactEvent(String json)
        el.$server.onReactEvent(JSON.stringify(msg));
    },
};

export {};
